File: src/minima_harness/tools/apply_patch.py
"""apply_patch — atomic multi-file, multi-hunk edits in one tool call.

A single ``edit`` does one ``old_string -> new_string`` swap, so a change spanning
three files costs three tool calls (three model round-trips) and can leave a
half-applied refactor if a later one fails. ``apply_patch`` takes one patch
describing adds/updates/deletes/moves across many files, resolves every hunk in
memory first, and only touches disk if *all* of them resolve — all-or-nothing.

The patch grammar mirrors the format Codex/OpenCode converged on::

    *** Begin Patch
    *** Add File: path/to/new.py
    +line one
    +line two
    *** Update File: path/to/existing.py
    *** Move to: path/to/renamed.py
    @@ optional anchor (e.g. a function signature in a big file)
     context line kept as-is
    -removed line
    +added line
    *** Delete File: path/to/old.py
    *** End Patch

Within an ``Update File`` hunk each line is prefixed by ``' '`` (context, kept),
``'-'`` (removed), or ``'+'`` (added). ``@@`` starts a new hunk and may carry an
anchor line to disambiguate location in large files. Matching tolerates trailing-
and (as a last resort) leading-whitespace drift so models needn't reproduce
indentation byte-for-byte.
"""

from __future__ import annotations

import difflib
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path

from pydantic import BaseModel

from minima_harness.agent.tools import AgentTool, ToolResult, error_result
from minima_harness.ai.types import TextContent

_BEGIN = "*** Begin Patch"
_END = "*** End Patch"
_ADD = "*** Add File: "
_UPDATE = "*** Update File: "
_DELETE = "*** Delete File: "
_MOVE = "*** Move to: "
_EOF = "*** End of File"

# Reader: relative-or-absolute path -> file text, or None if it doesn't exist.
ReadFile = Callable[[str], "str | None"]


class ApplyPatchParams(BaseModel):
    patch: str


class PatchError(ValueError):
    """A patch is malformed or cannot be applied cleanly to the current files."""


@dataclass(slots=True)
class Hunk:
    before: list[str]  # context + removed lines (prefix stripped), in file order
    after: list[str]  # context + added lines (prefix stripped), in file order
    anchor: str | None = None  # optional @@ heading to seek before matching


@dataclass(slots=True)
class FileChange:
    kind: str  # "add" | "update" | "delete"
    path: str
    move_to: str | None = None
    new_content: str | None = None  # for "add"
    hunks: list[Hunk] = field(default_factory=list)  # for "update"


@dataclass(slots=True)
class PatchPlan:
    writes: dict[str, str]  # path -> full new content
    deletes: list[str]  # paths to remove
    summary: list[str]  # one human-readable line per change


# --------------------------------------------------------------------------- parse


def parse_patch(text: str) -> list[FileChange]:
    """Parse patch text into file changes. Raises PatchError on malformed input."""
    lines = text.splitlines()
    i = 0
    while i < len(lines) and not lines[i].strip():
        i += 1
    if i >= len(lines) or lines[i].strip() != _BEGIN:
        raise PatchError("patch must start with '*** Begin Patch'")
    i += 1
    changes: list[FileChange] = []
    while i < len(lines):
        line = lines[i]
        s = line.strip()
        if s == _END:
            return changes
        if s == _EOF or not s:
            i += 1
            continue
        if line.startswith(_ADD):
            path = line[len(_ADD) :].strip()
            content, i = _parse_added_lines(lines, i + 1)
            changes.append(FileChange(kind="add", path=path, new_content=content))
            continue
        if line.startswith(_DELETE):
            changes.append(FileChange(kind="delete", path=line[len(_DELETE) :].strip()))
            i += 1
            continue
        if line.startswith(_UPDATE):
            path = line[len(_UPDATE) :].strip()
            i += 1
            move_to = None
            if i < len(lines) and lines[i].startswith(_MOVE):
                move_to = lines[i][len(_MOVE) :].strip()
                i += 1
            hunks, i = _parse_hunks(lines, i)
            if not hunks:
                raise PatchError(f"Update File: {path}: no hunks")
            changes.append(FileChange(kind="update", path=path, move_to=move_to, hunks=hunks))
            continue
        raise PatchError(f"unexpected line in patch: {line!r}")
    raise PatchError("patch missing '*** End Patch'")


def _parse_added_lines(lines: list[str], i: int) -> tuple[str, int]:
    out: list[str] = []
    while i < len(lines) and not lines[i].startswith("*** "):
        cl = lines[i]
        if cl.startswith("+"):
            out.append(cl[1:])
        elif not cl.strip():
            out.append("")
        else:
            raise PatchError(f"Add File lines must start with '+': {cl!r}")
        i += 1
    return "\n".join(out), i


def _parse_hunks(lines: list[str], i: int) -> tuple[list[Hunk], int]:
    hunks: list[Hunk] = []
    before: list[str] = []
    after: list[str] = []
    anchor: str | None = None

    def flush() -> None:
        nonlocal before, after, anchor
        if before or after:
            hunks.append(Hunk(before=before, after=after, anchor=anchor))
        before, after, anchor = [], [], None

    while i < len(lines):
        line = lines[i]
        if line.startswith("*** "):
            break
        if line.startswith("@@"):
            flush()
            anchor = line[2:].strip() or None
            i += 1
            continue
        if line.startswith("+"):
            after.append(line[1:])
        elif line.startswith("-"):
            before.append(line[1:])
        elif line.startswith(" "):
            before.append(line[1:])
            after.append(line[1:])
        elif line == "":  # blank context line emitted without the leading space
            before.append("")
            after.append("")
        else:
            raise PatchError(f"hunk line must start with ' ', '+', or '-': {line!r}")
        i += 1
    flush()
    return hunks, i


# --------------------------------------------------------------------------- plan


def plan_patch(changes: list[FileChange], read_file: ReadFile) -> PatchPlan:
    """Resolve every change against current file contents WITHOUT touching disk.

    Raises PatchError if any add collides, any delete/update target is missing, or
    any hunk fails to locate its context — so the caller can apply atomically."""
    writes: dict[str, str] = {}
    deletes: list[str] = []
    summary: list[str] = []
    for ch in changes:
        if ch.kind == "add":
            if read_file(ch.path) is not None:
                raise PatchError(f"Add File: {ch.path} already exists")
            content = ch.new_content or ""
            if content and not content.endswith("\n"):
                content += "\n"
            writes[ch.path] = content
            summary.append(f"add    {ch.path}")
        elif ch.kind == "delete":
            if read_file(ch.path) is None:
                raise PatchError(f"Delete File: {ch.path} does not exist")
            deletes.append(ch.path)
            summary.append(f"delete {ch.path}")
        elif ch.kind == "update":
            original = read_file(ch.path)
            if original is None:
                raise PatchError(f"Update File: {ch.path} does not exist")
            new_text = _apply_hunks(original, ch.hunks, ch.path)
            dest = ch.move_to or ch.path
            if ch.move_to:
                deletes.append(ch.path)
                summary.append(f"move   {ch.path} -> {ch.move_to}")
            else:
                summary.append(f"update {ch.path}")
            writes[dest] = new_text
        else:  # pragma: no cover - parse never produces other kinds
            raise PatchError(f"unknown change kind: {ch.kind}")
    return PatchPlan(writes=writes, deletes=deletes, summary=summary)


def _apply_hunks(original: str, hunks: list[Hunk], path: str) -> str:
    had_final_nl = original.endswith("\n")
    out = original.splitlines()
    cursor = 0
    for h in hunks:
        start = cursor
        if h.anchor:
            a = _find_anchor(out, h.anchor, cursor)
            if a >= 0:
                start = a
        idx = _find(out, h.before, start)
        if idx < 0 and start > 0:  # context may sit before the cursor — retry from top
            idx = _find(out, h.before, 0)
        if idx < 0:
            ctx = "\n".join(h.before[:6]) or "(no context lines)"
            raise PatchError(f"Update File: {path}: could not locate hunk context:\n{ctx}")
        out[idx : idx + len(h.before)] = h.after
        cursor = idx + len(h.after)
    text = "\n".join(out)
    if had_final_nl and not text.endswith("\n"):
        text += "\n"
    return text


def _find(hay: list[str], needle: list[str], start: int) -> int:
    """First index >= start where needle matches, trying progressively looser
    whitespace normalization. Empty needle matches at start (pure insertion)."""
    if not needle:
        return start
    span = len(needle)
    last = len(hay) - span
    for norm in (lambda x: x, str.rstrip, str.strip):
        target = [norm(x) for x in needle]
        for i in range(max(start, 0), last + 1):
            if [norm(x) for x in hay[i : i + span]] == target:
                return i
    return -1


def _find_anchor(hay: list[str], anchor: str, start: int) -> int:
    target = anchor.strip()
    for i in range(max(start, 0), len(hay)):
        if hay[i].strip() == target:
            return i
    return -1


# --------------------------------------------------------------------------- apply


def disk_reader(root: Path) -> ReadFile:
    """A ReadFile that resolves relative paths under ``root`` and reads from disk."""

    def read(rel: str) -> str | None:
        p = Path(rel) if Path(rel).is_absolute() else root / rel
        try:
            return p.expanduser().read_text(encoding="utf-8")
        except (FileNotFoundError, NotADirectoryError, IsADirectoryError):
            return None

    return read


def write_plan(plan: PatchPlan, root: Path) -> None:
    """Flush a resolved plan to disk: writes (with mkdir) first, then deletes."""
    for rel, content in plan.writes.items():
        p = Path(rel) if Path(rel).is_absolute() else root / rel
        p = p.expanduser()
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(content, encoding="utf-8")
    for rel in plan.deletes:
        if rel in plan.writes:  # a moved-to-itself or re-created path; keep it
            continue
        p = Path(rel) if Path(rel).is_absolute() else root / rel
        p.expanduser().unlink(missing_ok=True)


def patch_preview(patch: str, root: Path) -> str:
    """A unified-diff preview of what the patch would do, for the approval modal."""
    try:
        plan = plan_patch(parse_patch(patch), disk_reader(root))
    except PatchError as e:
        return f"apply_patch: {e}\n---\n{patch}"
    read = disk_reader(root)
    blocks: list[str] = []
    for rel, content in plan.writes.items():
        before = read(rel)
        d = difflib.unified_diff(
            (before or "").splitlines(),
            content.splitlines(),
            fromfile="/dev/null" if before is None else f"a/{rel}",
            tofile=f"b/{rel}",
            lineterm="",
        )
        blocks.append("\n".join(d) or f"{rel} (no change)")
    for rel in plan.deletes:
        if rel not in plan.writes:
            blocks.append(f"--- a/{rel}\n+++ /dev/null (deleted)")
    return "\n".join(blocks) or "apply_patch: empty patch"


def summarize_patch(patch: str) -> str:
    """A compact one-block summary of the patch for the transcript."""
    try:
        changes = parse_patch(patch)
    except PatchError as e:
        return f"apply_patch: {e}"
    rows: list[str] = []
    for ch in changes:
        if ch.kind == "add":
            rows.append(f"add    {ch.path}")
        elif ch.kind == "delete":
            rows.append(f"delete {ch.path}")
        elif ch.move_to:
            rows.append(f"move   {ch.path} -> {ch.move_to}")
        else:
            plural = "s" if len(ch.hunks) != 1 else ""
            rows.append(f"update {ch.path}  ({len(ch.hunks)} hunk{plural})")
    n = len(rows)
    head = f"apply_patch: {n} file{'s' if n != 1 else ''}"
    return head + "\n" + "\n".join(rows) if rows else head


async def _execute(tool_call_id: str, params, signal, on_update) -> ToolResult:  # noqa: ANN001
    assert isinstance(params, ApplyPatchParams)
    root = Path.cwd()
    try:
        changes = parse_patch(params.patch)
        if not changes:
            return error_result("apply_patch: empty patch (no file sections)")
        plan = plan_patch(changes, disk_reader(root))
    except PatchError as e:
        return error_result(f"apply_patch: {e}")
    write_plan(plan, root)
    summary = "\n".join(plan.summary)
    return ToolResult(
        content=[TextContent(text=f"applied patch ({len(changes)} change(s)):\n{summary}")],
        details={"writes": list(plan.writes), "deletes": plan.deletes},
    )


def apply_patch_tool() -> AgentTool:
    return AgentTool(
        name="apply_patch",
        description=(
            "Apply a multi-file patch atomically in one call — add, update, delete, or move "
            "files together. Prefer this over multiple `edit` calls for any change touching "
            "more than one file or more than one region. Every hunk is resolved before "
            "anything is written; if any hunk fails to match, NO file is changed.\n\n"
            "Format:\n"
            "*** Begin Patch\n"
            "*** Add File: path/new.py\n"
            "+full contents, each line prefixed with +\n"
            "*** Update File: path/existing.py\n"
            "*** Move to: path/renamed.py   (optional; omit to edit in place)\n"
            "@@ optional anchor line for large files\n"
            " unchanged context line (leading space)\n"
            "-removed line\n"
            "+added line\n"
            "*** Delete File: path/old.py\n"
            "*** End Patch\n\n"
            "Include a few unchanged context lines (leading space) around each change so the "
            "hunk can be located. Paths are relative to the working directory."
        ),
        parameters=ApplyPatchParams,
        execute=_execute,
    )

File: packages/tui/src/tools/apply_patch.ts
/**
 * apply_patch — atomic multi-file, multi-hunk edits in one tool call.
 * Port of minima_harness/tools/apply_patch.py.
 *
 * A single `edit` does one old→new swap, so a change spanning three files costs
 * three tool calls and can leave a half-applied refactor if a later one fails.
 * `apply_patch` takes one patch describing adds/updates/deletes/moves across many
 * files, resolves every hunk in memory first, and only touches disk if ALL of them
 * resolve — all-or-nothing.
 *
 * Grammar (Codex/OpenCode format):
 *
 *   *** Begin Patch
 *   *** Add File: path/new.ts
 *   +line one
 *   *** Update File: path/existing.ts
 *   *** Move to: path/renamed.ts        (optional)
 *   @@ optional anchor
 *    context line kept as-is
 *   -removed line
 *   +added line
 *   *** Delete File: path/old.ts
 *   *** End Patch
 */

import { mkdir, readFile, rm, writeFile } from "node:fs/promises";
import { dirname } from "node:path";
import { type AgentTool, type ToolResult, errorResult } from "../agent/tools.ts";
import { text } from "../ai/types.ts";
import { resolveWithin } from "./_io.ts";
import { objectSchema } from "./schema.ts";
import type { FsToolOptions } from "./types.ts";

const BEGIN = "*** Begin Patch";
const END = "*** End Patch";
const ADD = "*** Add File: ";
const UPDATE = "*** Update File: ";
const DELETE = "*** Delete File: ";
const MOVE = "*** Move to: ";
const EOF = "*** End of File";

/** A patch is malformed or cannot be applied cleanly to the current files. */
export class PatchError extends Error {}

interface Hunk {
  before: string[]; // context + removed lines (prefix stripped), in file order
  after: string[]; // context + added lines (prefix stripped), in file order
  anchor: string | null; // optional @@ heading to seek before matching
}

interface FileChange {
  kind: "add" | "update" | "delete";
  path: string;
  moveTo?: string | null;
  newContent?: string | null; // for "add"
  hunks: Hunk[]; // for "update"
}

export interface PatchPlan {
  writes: Map<string, string>; // path -> full new content
  deletes: string[]; // paths to remove
  summary: string[]; // one human-readable line per change
}

/** relative-or-absolute path -> file text, or null if it doesn't exist. */
type ReadFile = (path: string) => string | null;

// --------------------------------------------------------------------------- parse

export function parsePatch(patch: string): FileChange[] {
  const lines = patch.split("\n");
  let i = 0;
  while (i < lines.length && !lines[i]!.trim()) i++;
  if (i >= lines.length || lines[i]!.trim() !== BEGIN) {
    throw new PatchError("patch must start with '*** Begin Patch'");
  }
  i++;
  const changes: FileChange[] = [];
  while (i < lines.length) {
    const line = lines[i]!;
    const s = line.trim();
    if (s === END) return changes;
    if (s === EOF || !s) {
      i++;
      continue;
    }
    if (line.startsWith(ADD)) {
      const path = line.slice(ADD.length).trim();
      const [content, next] = parseAddedLines(lines, i + 1);
      changes.push({ kind: "add", path, newContent: content, hunks: [] });
      i = next;
      continue;
    }
    if (line.startsWith(DELETE)) {
      changes.push({ kind: "delete", path: line.slice(DELETE.length).trim(), hunks: [] });
      i++;
      continue;
    }
    if (line.startsWith(UPDATE)) {
      const path = line.slice(UPDATE.length).trim();
      i++;
      let moveTo: string | null = null;
      if (i < lines.length && lines[i]!.startsWith(MOVE)) {
        moveTo = lines[i]!.slice(MOVE.length).trim();
        i++;
      }
      const [hunks, next] = parseHunks(lines, i);
      if (!hunks.length) throw new PatchError(`Update File: ${path}: no hunks`);
      changes.push({ kind: "update", path, moveTo, hunks });
      i = next;
      continue;
    }
    throw new PatchError(`unexpected line in patch: ${JSON.stringify(line)}`);
  }
  throw new PatchError("patch missing '*** End Patch'");
}

function parseAddedLines(lines: string[], start: number): [string, number] {
  const out: string[] = [];
  let i = start;
  while (i < lines.length && !lines[i]!.startsWith("*** ")) {
    const cl = lines[i]!;
    if (cl.startsWith("+")) out.push(cl.slice(1));
    else if (!cl.trim()) out.push("");
    else throw new PatchError(`Add File lines must start with '+': ${JSON.stringify(cl)}`);
    i++;
  }
  return [out.join("\n"), i];
}

function parseHunks(lines: string[], start: number): [Hunk[], number] {
  const hunks: Hunk[] = [];
  let i = start;
  let before: string[] = [];
  let after: string[] = [];
  let anchor: string | null = null;

  const flush = () => {
    if (before.length || after.length) hunks.push({ before, after, anchor });
    before = [];
    after = [];
    anchor = null;
  };

  while (i < lines.length) {
    const line = lines[i]!;
    if (line.startsWith("*** ")) break;
    if (line.startsWith("@@")) {
      flush();
      anchor = line.slice(2).trim() || null;
      i++;
      continue;
    }
    if (line.startsWith("+")) {
      after.push(line.slice(1));
    } else if (line.startsWith("-")) {
      before.push(line.slice(1));
    } else if (line.startsWith(" ")) {
      before.push(line.slice(1));
      after.push(line.slice(1));
    } else if (line === "") {
      before.push("");
      after.push("");
    } else {
      throw new PatchError(`hunk line must start with ' ', '+', or '-': ${JSON.stringify(line)}`);
    }
    i++;
  }
  flush();
  return [hunks, i];
}

// --------------------------------------------------------------------------- plan

export function planPatch(changes: FileChange[], readFile: ReadFile): PatchPlan {
  const writes = new Map<string, string>();
  const deletes: string[] = [];
  const summary: string[] = [];
  for (const ch of changes) {
    if (ch.kind === "add") {
      if (readFile(ch.path) !== null) throw new PatchError(`Add File: ${ch.path} already exists`);
      let content = ch.newContent ?? "";
      if (content && !content.endsWith("\n")) content += "\n";
      writes.set(ch.path, content);
      summary.push(`add    ${ch.path}`);
    } else if (ch.kind === "delete") {
      if (readFile(ch.path) === null)
        throw new PatchError(`Delete File: ${ch.path} does not exist`);
      deletes.push(ch.path);
      summary.push(`delete ${ch.path}`);
    } else {
      const original = readFile(ch.path);
      if (original === null) throw new PatchError(`Update File: ${ch.path} does not exist`);
      const newText = applyHunks(original, ch.hunks, ch.path);
      const dest = ch.moveTo || ch.path;
      if (ch.moveTo) {
        deletes.push(ch.path);
        summary.push(`move   ${ch.path} -> ${ch.moveTo}`);
      } else {
        summary.push(`update ${ch.path}`);
      }
      writes.set(dest, newText);
    }
  }
  return { writes, deletes, summary };
}

function applyHunks(original: string, hunks: Hunk[], path: string): string {
  const hadFinalNl = original.endsWith("\n");
  const out = original.split("\n");
  // A trailing newline produces a spurious empty final element; drop it so line
  // indices match the file's real lines (re-added below if present).
  if (hadFinalNl && out.length && out[out.length - 1] === "") out.pop();
  let cursor = 0;
  for (const h of hunks) {
    let start = cursor;
    if (h.anchor) {
      const a = findAnchor(out, h.anchor, cursor);
      if (a >= 0) start = a;
    }
    let idx = find(out, h.before, start);
    if (idx < 0 && start > 0) idx = find(out, h.before, 0); // context may sit before the cursor
    if (idx < 0) {
      const ctx = h.before.slice(0, 6).join("\n") || "(no context lines)";
      throw new PatchError(`Update File: ${path}: could not locate hunk context:\n${ctx}`);
    }
    out.splice(idx, h.before.length, ...h.after);
    cursor = idx + h.after.length;
  }
  let out_text = out.join("\n");
  if (hadFinalNl && !out_text.endsWith("\n")) out_text += "\n";
  return out_text;
}

/**
 * First index >= start where needle matches, trying progressively looser
 * whitespace normalization. Empty needle matches at start (pure insertion).
 */
function find(hay: string[], needle: string[], start: number): number {
  if (!needle.length) return start;
  const span = needle.length;
  const last = hay.length - span;
  const norms: ((x: string) => string)[] = [
    (x) => x,
    (x) => x.replace(/\s+$/, ""),
    (x) => x.trim(),
  ];
  for (const norm of norms) {
    const target = needle.map(norm);
    for (let i = Math.max(start, 0); i <= last; i++) {
      let ok = true;
      for (let j = 0; j < span; j++) {
        if (norm(hay[i + j]!) !== target[j]) {
          ok = false;
          break;
        }
      }
      if (ok) return i;
    }
  }
  return -1;
}

function findAnchor(hay: string[], anchor: string, start: number): number {
  const target = anchor.trim();
  for (let i = Math.max(start, 0); i < hay.length; i++) {
    if (hay[i]!.trim() === target) return i;
  }
  return -1;
}

// --------------------------------------------------------------------------- apply

function makeResolver(workdir?: string): (rel: string) => string {
  return (rel: string) => {
    const r = resolveWithin(rel, workdir);
    if (!r.ok) throw new PatchError(r.error);
    return r.path;
  };
}

/**
 * Pre-read every file the parsed changes reference so planPatch can stay sync.
 * Mirrors the old disk reader exactly: unreadable or missing files map to null
 * (an Add File target need not exist), and a path-resolution failure is stored
 * and re-thrown only when planPatch actually reads that path.
 */
async function preloadFiles(
  changes: FileChange[],
  resolve: (rel: string) => string,
): Promise<Map<string, string | null | PatchError>> {
  const files = new Map<string, string | null | PatchError>();
  for (const ch of changes) {
    if (files.has(ch.path)) continue;
    let abs: string;
    try {
      abs = resolve(ch.path);
    } catch (exc) {
      if (exc instanceof PatchError) {
        files.set(ch.path, exc);
        continue;
      }
      throw exc;
    }
    try {
      files.set(ch.path, await readFile(abs, "utf8"));
    } catch {
      files.set(ch.path, null);
    }
  }
  return files;
}

function mapReader(files: Map<string, string | null | PatchError>): ReadFile {
  return (rel: string) => {
    const cached = files.get(rel);
    if (cached instanceof PatchError) throw cached;
    return cached ?? null;
  };
}

/** Flush a resolved plan to disk: writes (with mkdir) first, then deletes. */
export async function writePlan(plan: PatchPlan, resolve: (rel: string) => string): Promise<void> {
  for (const [rel, content] of plan.writes) {
    const abs = resolve(rel);
    await mkdir(dirname(abs), { recursive: true });
    await writeFile(abs, content, "utf8");
  }
  for (const rel of plan.deletes) {
    if (plan.writes.has(rel)) continue; // a moved-to-itself or re-created path; keep it
    await rm(resolve(rel), { force: true });
  }
}

const parameters = objectSchema({ patch: { type: "string", description: "The patch text." } }, [
  "patch",
]);

export function applyPatchTool(opts: FsToolOptions = {}): AgentTool {
  return {
    name: "apply_patch",
    description:
      "Apply a multi-file patch atomically in one call — add, update, delete, or move " +
      "files together. Prefer this over multiple `edit` calls for any change touching " +
      "more than one file or more than one region. Every hunk is resolved before " +
      "anything is written; if any hunk fails to match, NO file is changed.\n\n" +
      "Format:\n" +
      "*** Begin Patch\n" +
      "*** Add File: path/new.ts\n" +
      "+full contents, each line prefixed with +\n" +
      "*** Update File: path/existing.ts\n" +
      "*** Move to: path/renamed.ts   (optional; omit to edit in place)\n" +
      "@@ optional anchor line for large files\n" +
      " unchanged context line (leading space)\n" +
      "-removed line\n" +
      "+added line\n" +
      "*** Delete File: path/old.ts\n" +
      "*** End Patch\n\n" +
      "Include a few unchanged context lines (leading space) around each change so the " +
      "hunk can be located. Paths are relative to the working directory.",
    parameters,
    async execute(_id: string, params: Record<string, unknown>): Promise<ToolResult> {
      const resolve = makeResolver(opts.workdir);
      let plan: PatchPlan;
      let changeCount: number;
      try {
        const changes = parsePatch(String(params.patch));
        if (!changes.length) return errorResult("apply_patch: empty patch (no file sections)");
        changeCount = changes.length;
        const files = await preloadFiles(changes, resolve);
        plan = planPatch(changes, mapReader(files));
      } catch (exc) {
        if (exc instanceof PatchError) return errorResult(`apply_patch: ${exc.message}`);
        throw exc;
      }
      await writePlan(plan, resolve);
      const summary = plan.summary.join("\n");
      return {
        content: [text(`applied patch (${changeCount} change(s)):\n${summary}`)],
        details: { writes: [...plan.writes.keys()], deletes: plan.deletes },
      };
    },
  };
}
